recipe-client-addon/lib/PreferenceExperiments.jsm

```diff
@@ -0,0 +1,296 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Preference Experiments temporarily change a preference to one of several test
+ * values for the duration of the experiment. Telemetry packets are annotated to
+ * show what experiments are active, and we use this data to measure the
+ * effectiveness of the preference change.
+ *
+ * Info on active and past experiments is stored in a JSON file in the profile
+ * folder.
+ *
+ * Active preference experiments are stopped if they aren't active on the recipe
+ * server. They also expire if Firefox isn't able to contact the recipe server
+ * after a period of time, as well as if the user modifies the preference during
+ * an active experiment.
+ */
+
+/**
+ * Experiments store info about an active or expired preference experiment.
+ * They are single-depth objects to simplify cloning.
+ * @typedef {Object} Experiment
+ * @property {string} name
+ *   Unique name of the experiment
+ * @property {string} branch
+ *   Experiment branch that the user was matched to
+ * @property {boolean} expired
+ *   If false, the experiment is active.
+ * @property {string} lastSeen
+ *   ISO-formatted date string of when the experiment was last seen from the
+ *   recipe server.
+ * @property {string} preferenceName
+ *   Name of the preference affected by this experiment.
+ * @property {string|integer|boolean} preferenceValue
+ *   Value to change the preference to during the experiment.
+ * @property {string|integer|boolean|undefined} previousPreferenceValue
+ *   Value of the preference prior to the experiment, or undefined if it was
+ *   unset.
+ */
+
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "CleanupManager", "resource://shield-recipe-client/lib/CleanupManager.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "JSONFile", "resource://gre/modules/JSONFile.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "OS", "resource://gre/modules/osfile.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "LogManager", "resource://shield-recipe-client/lib/LogManager.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "Preferences", "resource://gre/modules/Preferences.jsm");
+
+this.EXPORTED_SYMBOLS = ["PreferenceExperiments"];
+
+const EXPERIMENT_FILE = "shield-preference-experiments.json";
+
+/**
+ * Asynchronously load the JSON file that stores experiment status in the profile.
+ */
+let storePromise;
+function loadStorage() {
```

Comment:
Nit: because this doesn't always actually do a file load, I think calling it `ensureStorage` or similar would make more sense.